applications/mne_sample_set_downloader/downloader.cpp

```diff
@@ -0,0 +1,216 @@
+//=============================================================================================================
+// QT INCLUDES
+//=============================================================================================================
+
+#include <QFileDialog>
+
+//*************************************************************************************************************
+//=============================================================================================================
+// DEFINE MEMBER METHODS
+//=============================================================================================================
+
+#ifdef _WIN32
+
+void Downloader::on_extractButton_clicked()
+{
+    QString zipPath = "\"" + QString(QDir::toNativeSeparators(ui->filePath->text())) + "\"";
+    readyToExtract(boolDownloadStatus, zipPath, m_qCurrentPath);
+    ui->extractButton->setEnabled(false);
+
+}
+
+void Downloader::checkZipperPath()
+{
+    ui->label->setText("7zip installed? Please check the path");
+    ui->extractButton->setVisible(true);
+    ui->extractButton->setEnabled(true);
+    ui->downloadButton->setVisible(false);
+    ui->downloadButton->setEnabled(false);
+    ui->progressBar->setVisible(false);
+    ui->filePath->setVisible(true);
+    ui->toolButton->setVisible(true);
+    ui->toolButton->setEnabled(true);
+}
+
+void Downloader::on_toolButton_clicked()
+{
+    QString dir = QFileDialog::getExistingDirectory(this, tr("Open Directory"), "/home",QFileDialog::DontResolveSymlinks) + "\\7z.exe";
+    ui->filePath->setText(QDir::toNativeSeparators(dir));
+}
+
+void Downloader::readyToExtract(bool stat, QString zip, QString current)
+{
+    if (stat == true)
+    {
+        ui->label->setText("EXTRACTING...");
+        ui->progressBar->setVisible(false);
+        QObject::connect(&m_extractor, SIGNAL(zipperError()), this, SLOT(checkZipperPath()));
+        connect(&m_extractor, SIGNAL(extractionDone()), this, SLOT(done()));
+        m_extractor.beginExtraction(zip, current);
+    }
+}
+
+void Downloader::downloadFinished()
+{
+    if (m_qReply != NULL)
+        m_qFile.write(m_qReply->readAll());
+    m_qFile.close();
+    boolDownloadStatus = true;
+    QString zipPath ="\"" + QString (ui->filePath->text()) + "\"";
+    ui->progressBar->setEnabled(false);
+    readyToExtract(boolDownloadStatus, zipPath, m_qCurrentPath);
+    return;
+}
+#elif __linux__
+
+void Downloader::readyToExtract(bool stat)
+{
+    if (stat == true)
+    {
+        ui->label->setText("EXTRACTING...");
+        m_extractor.beginExtraction();
+    }
+}
+
+void Downloader::downloadFinished()
+{
+    if (m_qReply != NULL)
+        m_qFile.write(m_qReply->readAll());
+    boolDownloadStatus = true;
+    ui->progressBar->setEnabled(false);
+    readyToExtract(boolDownloadStatus);
+    return;
+}
+
+#endif
+
+void Downloader::done()
+{
+    ui->label->setText("done");
+}
+
+//placeholder downloadFinished OSX
```

Comment:
Change this to TODO: Add OSX support. QtCreator can conveniently scan for TODO strings in our project files. This way we can always generate a To Do list from our code when using the TODO syntax